river/rules/base.py

```diff
@@ -0,0 +1,275 @@
+            )
+
+            if b_split.merit > 0 and (b_split.merit - sb_split.merit > hb or hb < tau):
+                should_expand = True
+
+        if should_expand:
+            b_split = suggestions[-1]
+            is_numerical = b_split.numerical_feature
+            branch_no = self.split_criterion.select_best_branch(b_split.children_stats)
+
+            if is_numerical:
+                lit = NumericLiteral(
+                    b_split.feature, b_split.split_info, branch_no != 0
+                )
+
+                literal_updated = False
+                for literal in self.literals:
+                    if lit.on == literal.on and lit.neg == literal.neg:
+                        # Update thresholds rather than adding a new literal
+                        if not literal.neg and lit.at < literal.at:
+                            literal.at = lit.at
+                            literal_updated = True
+                            break
+                        elif literal.neg and lit.at > literal.at:
+                            literal.at = lit.at
+                            literal_updated = True
+                            break
+
+                # No threshold was updated, thus a new literal is added
+                if not literal_updated:
+                    self.literals.append(lit)
+            else:
+                lit = NominalLiteral(
+                    b_split.feature, b_split.split_info, branch_no != 0
+                )
+                # Add a new literal
+                self.literals.append(lit)
+
+            # Reset all the statistics stored in the the decision rule
+            updated_rule = self.clone()
+            # Keep the literals
+            updated_rule.literals.extend(self.literals)
+
+            return updated_rule, True
+
+        # If the expansion attempt failed, update the expansion tracker
+        self._last_expansion_attempt_at = self.total_weight
+        return self, False
+
+    def covers(self, x: dict) -> bool:
+        """Check if all the conditions are fulfilled.
+
+        Parameters
+        ----------
+        x
+            Input instance.
+        """
+        return all(map(lambda lit: lit(x), self.literals))
+
+    @abc.abstractmethod
+    def new_nominal_splitter(self):
+        pass
+
+    def _iter_features(self, x: dict) -> typing.Iterable:
+        """Determine how the input instance is looped through when updating the splitters.
+
+        Parameters
+        ----------
+        x
+            The input instance.
+        """
+        for att_id, att_val in x.items():
+            yield att_id, att_val
+
+    @abc.abstractmethod
+    def _update_target_stats(self, y, w):
+        pass
+
+    def _update_feature_stats(self, feat_name, feat_val, w):
+        # By default, no feature stats are kept
+        pass
+
+    def update(self, x, y, w):
+        self._total_weight += w
+        self._update_target_stats(y, w)
```

Comment:
is this necessary?

Reply:
If I'm not wrong, I had to keep a separate estimate of the observed weight because the target (split) stats are reset after each expansion. Hence, `total_weight` cannot be obtained reliably from the target stats.

Although the newly expanded rules are not ready to be expanded again yet, they can perform anomaly detection.

AMRules' rules have a warm start before being used to detect anomalies. `total_weight` is checked for this end.

Reply:
Sorry, I was not clear. `self._update_target_stats` is not implemented. Is it implemented somewhere else?

Reply:
Sorry, I was not clear. `self._update_target_stats` is not implemented. Is it implemented somewhere else?